lib/ae_mdw/mnesia.ex

```diff
@@ -0,0 +1,139 @@
+defmodule AeMdw.Mnesia do
+  @moduledoc """
+  Mnesia wrapper to provide a simpler API.
+
+  In order to iterate throught collections of records we use the
+  concept of "cursor". A cursor is just a key of any given table that
+  is returned in order the access the next page of records.
+
+  This cursor key is whatever the current key of the table might be
+  (e.g. a tuple, a number). If there is no next page, `nil` is
+  returned instead.
+  """
+
+  @type table() :: atom()
+  @type record() :: tuple()
+  @type key() :: term()
+  @type direction() :: :forward | :backward
+  @type cursor() :: key() | nil
+
+  @end_token :"$end_of_table"
+
+  @spec last_key(table()) :: {:ok, key()} | :none
+  def last_key(tab) do
+    case :mnesia.dirty_last(tab) do
+      @end_token -> :none
+      last_key -> {:ok, last_key}
+    end
+  end
+
+  @spec first_key(table()) :: {:ok, key()} | :none
+  def first_key(tab) do
+    case :mnesia.dirty_first(tab) do
+      @end_token -> :none
+      last_key -> {:ok, last_key}
+    end
+  end
+
+  @spec fetch(table(), key()) :: {:ok, record()} | :not_found
+  def fetch(tab, key) do
+    case :mnesia.dirty_read(tab, key) do
+      [record] -> {:ok, record}
+      [] -> :not_found
+    end
+  end
+
+  @spec fetch!(table(), key()) :: record()
+  def fetch!(tab, key) do
+    {:ok, record} = fetch(tab, key)
+
+    record
+  end
+
+  @spec fetch_keys(table(), direction(), cursor(), non_neg_integer()) :: {[record()], cursor()}
+  def fetch_keys(tab, :forward, nil, limit) do
+    {:atomic, {keys, cursor}} =
+      :mnesia.transaction(fn ->
```

Comment:
Took off the hat for the solution of nested contexts. Nice use of `Stream.unfold` :clap: 
Only here the mnesia context could continue to be dirty for lighter fetching with `async_dirty` like in `AeMdw.Db.Util`.

Reply:
Done. With it I reduced the memory usage, but increased the time just a slightly bit. But I think async_dirty/1 is better anyway, nice catch!

![Screenshot from 2021-09-10 21-09-22](https://user-images.githubusercontent.com/750314/132929319-5861aa1d-0f95-4a4f-86c7-66594346fdae.png)

Reply:
Wondering if `:mnesia.sync_dirty` would be faster for reading (empty log to write) as the `:mnesia.transaction` was faster than `:mnesia.async_dirty`.

Reply:
Using `:mnesia.sync_dirty/1` reports exactly as much memory as before and doesn't really change much in terms of timing. So I think I'm leaving `async_dirty`